java/src/main/java/ai/onnxruntime/OnnxRuntime.java

```diff
@@ -5,175 +5,116 @@
+      logger.log(Level.FINE, "Attempting to load native library '" + library + "' from specified path: " + libraryPathProperty);
+      File libraryFile = new File(libraryPathProperty);
+      String libraryFilePath = libraryFile.getAbsolutePath();
+      if (!libraryFile.exists()) {
+        throw new IOException("Native library '" + library + "' not found at "+ libraryFilePath);
+      }
+      System.load(libraryFilePath);
+      logger.log(Level.FINE, "Loaded native library '" + library + "' from specified path");
+      return;
+    }
 
-            if (filename != null && prefix.length() >= 3) {
-                File temp = File.createTempFile(prefix, suffix);
-                if (debugLogging) {
-                    logger.info("Writing " + path + " out to " + temp.getAbsolutePath());
-                }
-                temp.deleteOnExit();
-                if (!temp.exists()) {
-                    throw new FileNotFoundException("File " + temp.getAbsolutePath() + " does not exist.");
-                } else {
-                    byte[] buffer = new byte[1024];
-                    try (InputStream is = OnnxRuntime.class.getResourceAsStream(path)) {
-                        if (is == null) {
-                            throw new FileNotFoundException("File " + path + " was not found inside JAR.");
-                        } else {
-                            int readBytes;
-                            try (FileOutputStream os = new FileOutputStream(temp)) {
-                                while ((readBytes = is.read(buffer)) != -1) {
-                                    os.write(buffer, 0, readBytes);
-                                }
-                            }
-                            return temp.getAbsolutePath();
-                        }
-                    }
-                }
-            } else {
-                throw new IllegalArgumentException("The filename has to be at least 3 characters long.");
-            }
+    // 3) try loading from resources or library path:
+    // generate a platform specific library name
+    // replace Mac's jnilib extension to dylib
+    String libraryFileName = System.mapLibraryName(library).replace("jnilib", "dylib");
+    String resourcePath = "/ai/onnxruntime/native/" + libraryFileName;
+    InputStream is = OnnxRuntime.class.getResourceAsStream(resourcePath);
+    if (is == null) {
+      // 3a) Not found in resources, load from library path
+      logger.log(Level.FINE, "Attempting to load native library '" + library + "' from library path");
+      System.loadLibrary(library);
+      logger.log(Level.FINE, "Loaded native library '" + library + "' from library path");
+    } else {
+      // 3b) Found in resources, load via temporary file
+      logger.log(
+          Level.FINE,
+          "Attempting to load native library '" + library + "' from resource path " + resourcePath);
+      File temp = File.createTempFile("javaload_", libraryFileName);
```

Comment:
Looking at #2892 this might cause an issue on Windows. I'll try to test out my proposed fix this week if I can persuade my desktop to have the right Windows build environment (once I've checked it doesn't change macOS or Linux), and then we'd need to incorporate that change into this loader.

Reply:
OK sounds good, since I do not have a windows environment.

Do we think this is the reason the windows CI failed? @snnn should we rerun that part?

Taking a look at the netty loader: 
https://github.com/netty/netty/blob/4.1/common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java#L183

They set the prefix and suffix such that `onnxruntime.dll` becomes `onnxruntime123452.dll` maybe the prefix needs to match like that for windows to play nicely. However your original work seems to do this already. Additionally, they have a flush() call. Maybe that is something that windows really likes. I have no other ideas at the moment.

Reply:
I have a fix for #2892, https://github.com/Craigacp/onnxruntime/commit/3aef28097a53cdd8bfa9a0a05ed1106a30acef20. Do you want to mirror the logic into this PR? Or I can submit a PR to your branch with the fix.

Reply:
I'll mirror it here shortly